homeassistant/components/onboarding/__init__.py

```diff
@@ -50,21 +67,37 @@ async def _async_migrate_func(
 @callback
 def async_is_onboarded(hass: HomeAssistant) -> bool:
     """Return if Home Assistant has been onboarded."""
-    data = hass.data.get(DOMAIN)
-    return data is None or data is True
+    data: OnboardingData | None = hass.data.get(DOMAIN)
+    return data is None or data.onboarded is True
 
 
 @bind_hass
 @callback
 def async_is_user_onboarded(hass: HomeAssistant) -> bool:
     """Return if a user has been created as part of onboarding."""
-    return async_is_onboarded(hass) or STEP_USER in hass.data[DOMAIN]["done"]
+    return async_is_onboarded(hass) or STEP_USER in hass.data[DOMAIN].steps["done"]
+
+
+@callback
+def async_add_listener(hass: HomeAssistant, listener: Callable[[], None]) -> None:
+    """Add a listener to be called when onboarding is complete."""
+    data: OnboardingData | None = hass.data.get(DOMAIN)
+
+    if not data:
+        # Onboarding not active
+        return
+
+    if data.onboarded:
+        listener()
+        return
```

Comment:
This feels a bit weird.

`if not data` happens when onboarding was finished while onboarding integration was set up, so we don't store it in `hass.data`. In this case we do not do anything.

`if data.onboarded` when the instance was not onboarded during startup but is onboarded now. In this case we call the listener directly. I don't think that this situation can happen, but it is weird that it's inconsistent with the previous check. In both cases we are already onboarded.